README.md
# Grafana Sequential Data Store Sample

**Version:** 1.2.2

[![Build Status](https://dev.azure.com/osieng/engineering/_apis/build/status/product-readiness/OCS/osisoft.sample-ocs-grafana-nodejs?repoName=osisoft%2Fsample-ocs-grafana-nodejs&branchName=main)](https://dev.azure.com/osieng/engineering/_build/latest?definitionId=2619&repoName=osisoft%2Fsample-ocs-grafana-nodejs&branchName=main)

This sample demonstrates how to build a [Grafana](https://grafana.com/) plugin that runs queries against the Sequential Data Store of OSIsoft Cloud Services or Edge Data Store. The sample performs normal "Get Values" calls against a specified stream in SDS, using the time range of the Grafana dashboard. See the [Grafana Documentation](https://grafana.com/docs/grafana/latest/developers/plugins/) for more information on developing Grafana plugins.

## Requirements

- [Grafana 7.0+](https://grafana.com/grafana/download)
- Web Browser with JavaScript enabled
- [NodeJS](https://nodejs.org/en/)
- [Git](https://git-scm.com/download/win)
- If using OSIsoft Cloud Services, register a Client Credentials Client in OSIsoft Cloud Services; a client secret will need to be provided to the sample plugin configuration
- If using Edge Data Store, the browser must be running local to a running copy of Edge Data Store

## Running the Sample with installed Grafana

1. Copy this folder to your Grafana server's plugins directory, like `.../grafana/data/plugins`
1. (Optional) If using other plugins, rename the folder to `osisoft-cloud-services-sample`
1. Open a command prompt inside that folder
1. Install dependencies, using `npm ci`
1. Build the plugin, using `npm run build` (or `npm run dev` for browser debugging)
1. Restart the Grafana server to load the new plugin
1. Add a new Grafana datasource using the sample (see [Grafana docs](https://grafana.com/docs/grafana/latest/features/datasources/add-a-data-source/))
1. Choose whether to query against OSIsoft Cloud Services or Edge Data Store
1. Enter the relevant required information; if using OCS, the client secret will be encrypted in the Grafana server and HTTP requests to OCS will be made by a server-side proxy, as described in the [Grafana docs](https://grafana.com/docs/grafana/latest/developers/plugins/authentication/)
1. Open a new or existing Grafana dashboard, and choose the Sequential Data Store Sample as the data source
1. Enter your Namespace (if querying OCS) and Stream, and data will populate into the dashboard from the stream for the dashboard's time range

## Running the Sample with Docker

1. Open a command prompt inside this folder
1. Build the container using `docker build -t grafana-ocs .`
1. Run the container using `docker run -d --name=grafana -p 3000:3000 grafana-ocs`
1. Navigate to localhost:3000 to configure data sources and view data

## Using OCS OAuth login to Grafana

A limitation of this sample is that the use of a Client Credentials Client means that anyone using the Data Source in Grafana receives the same access inside OSIsoft Cloud Services. Grafana supports generic OAuth login to the Grafana server, and can forward its token to the destination data source, including OCS. See [here](https://grafana.com/docs/grafana/latest/auth/generic-oauth/) for more information on this feature of Grafana. This feature could be used to ensure individual users must log in to receive the appropriate permissions in OCS.

However, Grafana's implementation of the Authorization Code Flow does not use [Proof of Key Code Exchange](https://oauth.net/2/pkce/), or PKCE, which is an additional security layer required by OCS Authorization Code Clients. Grafana also does not support OCS Hybrid Clients as it does not support a `POST` back from the authentication server, nor does it support the `response_type` and `response_mode` headers required by that flow.

If you are interested in using using OCS OAuth in your Grafana server, please check and comment on [this issue](https://github.com/grafana/grafana/issues/26350) in the Grafana GitHub repository.

## Using Community Data

1. Add a new Grafana datasource using the sample (see [Grafana docs](https://grafana.com/docs/grafana/latest/features/datasources/add-a-data-source/))
1. Choose OSIsoft Cloud Services
1. Toggle the "Community Data" switch to 'true'
1. Enter the relevant required information. You can find the Community ID in the URL of the Community Details page.

## Running the Automated Tests

1. Open a command prompt inside this folder
1. Install dependencies, using `npm ci`
1. Run the tests, using `npm test`

---

For the main OCS page [ReadMe](https://github.com/osisoft/OSI-Samples-OCS)
For the main samples page [ReadMe](https://github.com/osisoft/OSI-Samples)


HISTORY.md
# Version History

## 1.2.2 / 2021-07-15

- Updated dependencies

## 1.2.1 / 2021-07-06

- Updated dependencies

## 1.2.0 / 2021-06-01

- Added the ability to view data from OCS communities
- Added a dockerfile for containerized sample with Grafana server

## 1.1.7 / 2021-05-24

- Updated dependencies

## 1.1.6 / 2021-04-13

- Updated dependencies

## 1.1.5 / 2021-03-02

- Updated pipeline to use internal agent pool
- Updated pipeline to reference main branch

## 1.1.4 / 2021-01-28

- Fix linting issues

## 1.1.3 / 2021-01-26

- Updated dependencies

## 1.1.2 / 2020-11-02

- Updated dependencies

## 1.1.1 / 2020-09-14

- Updated dependencies

## 1.1.0 / 2020-08-11

- Converted HTML partials to React components
- Added namespace to Data Source configuration
- Added selection field for stream ID to query editor

## 1.0.2 / 2020-08-06

- Updated dependencies

## 1.0.1 / 2020-06-11

- Updated dependencies

## 1.0.0 / 2020-06-02

- Initial public release


src/DataSource.spec.ts
import { DataSourceInstanceSettings, MutableDataFrame, FieldType } from '@grafana/data';

import { SdsDataSource } from 'DataSource';
import { SdsDataSourceOptions, SdsDataSourceType } from 'types';

describe('SdsDataSource', () => {
  const url = 'URL';
  const edsPort = 'PORT';
  const ocsUrl = 'URL';
  const ocsVersion = 'VERSION';
  const ocsTenant = 'TENANT';
  const ocsClient = 'CLIENT';
  const oauthPassThru = false;
  const namespace = 'NAMESPACE';
  const ocsUseCommunity = false;
  const ocsCommunity = 'COMMUNITY';
  const ocsSettings: DataSourceInstanceSettings<SdsDataSourceOptions> = {
    id: 0,
    uid: '',
    name: '',
    type: '',
    url,
    meta: null as any,
    jsonData: {
      type: SdsDataSourceType.OCS,
      edsPort: edsPort,
      ocsUrl: ocsUrl,
      ocsVersion: ocsVersion,
      ocsTenant: ocsTenant,
      ocsClient: ocsClient,
      oauthPassThru,
      namespace,
      ocsUseCommunity: ocsUseCommunity,
      ocsCommunity: ocsCommunity,
    },
  };
  const ocsCommSettings = { ...ocsSettings, ...{ jsonData: { ...ocsSettings.jsonData, ocsUseCommunity: true } } };
  const edsSettings = { ...ocsSettings, ...{ jsonData: { ...ocsSettings.jsonData, type: SdsDataSourceType.EDS } } };
  const backendSrv = {
    datasourceRequest: () => new Promise((r) => r),
  };

  describe('constructor', () => {
    it('should use passed in data source information', () => {
      const datasource = new SdsDataSource(ocsSettings, backendSrv as any);
      expect(datasource.proxyUrl).toEqual(url);
      expect(datasource.type).toEqual(SdsDataSourceType.OCS);
      expect(datasource.edsPort).toEqual(edsPort);
      expect(datasource.ocsUrl).toEqual(ocsUrl);
      expect(datasource.ocsVersion).toEqual(ocsVersion);
      expect(datasource.ocsTenant).toEqual(ocsTenant);
      expect(datasource.ocsUseCommunity).toEqual(ocsUseCommunity);
      expect(datasource.ocsCommunity).toEqual(ocsCommunity);
      expect(datasource.oauthPassThru).toEqual(oauthPassThru);
      expect(datasource.namespace).toEqual(namespace);
    });

    it('should handle empty jsonData', () => {
      const nullSettings: any = { ...ocsSettings, ...{ url: null, jsonData: null } };
      const datasource = new SdsDataSource(nullSettings, backendSrv as any);
      expect(datasource.proxyUrl).toEqual('');
      expect(datasource.type).toEqual(SdsDataSourceType.OCS);
      expect(datasource.edsPort).toEqual('5590');
      expect(datasource.ocsUrl).toEqual('');
      expect(datasource.ocsVersion).toEqual('v1');
      expect(datasource.ocsTenant).toEqual('');
      expect(datasource.ocsUseCommunity).toEqual(false);
      expect(datasource.ocsCommunity).toEqual('');
      expect(datasource.oauthPassThru).toEqual(false);
      expect(datasource.namespace).toEqual('');
    });
  });

  describe('getStreamsUrl', () => {
    it('should return the correct URL for OCS', () => {
      const datasource = new SdsDataSource(ocsSettings, backendSrv as any);
      expect(datasource.streamsUrl).toEqual('URL/ocs/api/VERSION/tenants/TENANT/namespaces/NAMESPACE/streams');
    });

    it('should return the correct URL for OCS communities', () => {
      const datasource = new SdsDataSource(ocsCommSettings, backendSrv as any);
      expect(datasource.streamsUrl).toEqual('URL/ocs/api/VERSION/tenants/TENANT/search/communities/COMMUNITY/streams');
    });

    it('should return the correct URL for EDS', () => {
      const datasource = new SdsDataSource(edsSettings, backendSrv as any);
      expect(datasource.streamsUrl).toEqual(
        'http://localhost:PORT/api/v1/tenants/default/namespaces/NAMESPACE/streams'
      );
    });
  });

  describe('query', () => {
    it('should query with the expected parameters', (done) => {
      spyOn(backendSrv, 'datasourceRequest').and.returnValue(
        Promise.resolve({
          data: [
            {
              TimeStamp: '2020-01-01',
              Boolean: true,
              Number: 1,
              String: 'A',
            },
          ],
        })
      );
      const datasource = new SdsDataSource(ocsSettings, backendSrv as any);
      const options = {
        range: {
          from: {
            utc: () => ({
              format: () => 'FROM',
            }),
          },
          to: {
            utc: () => ({
              format: () => 'TO',
            }),
          },
        },
        targets: [
          {
            refId: 'REFID',
            namespace: 'NAMESPACE',
            streamId: 'STREAM',
            streamName: 'STREAM',
          },
        ],
      };
      const response = datasource.query(options as any);
      expect(backendSrv.datasourceRequest).toHaveBeenCalledWith({
        url: 'URL/ocs/api/VERSION/tenants/TENANT/namespaces/NAMESPACE/streams/STREAM/data?startIndex=FROM&endIndex=TO',
        method: 'GET',
      });
      response.then((r) => {
        expect(JSON.stringify(r)).toEqual(
          JSON.stringify({
            data: [
              new MutableDataFrame({
                refId: 'REFID',
                name: 'STREAM',
                fields: [
                  {
                    name: 'TimeStamp',
                    type: FieldType.time,
                    values: [Date.parse('2020-01-01')],
                  },
                  {
                    name: 'Boolean',
                    type: FieldType.boolean,
                    values: [true],
                  },
                  {
                    name: 'Number',
                    type: FieldType.number,
                    values: [1],
                  },
                  {
                    name: 'String',
                    type: FieldType.string,
                    values: ['A'],
                  },
                ],
              }),
            ],
          })
        );
        done();
      });
    });

    it('should query a community with the expected parameters', (done) => {
      spyOn(backendSrv, 'datasourceRequest').and.returnValue(
        Promise.resolve({
          data: [
            {
              TimeStamp: '2020-01-01',
              Boolean: true,
              Number: 1,
              String: 'A',
            },
          ],
        })
      );
      const datasource = new SdsDataSource(ocsCommSettings, backendSrv as any);
      const options = {
        range: {
          from: {
            utc: () => ({
              format: () => 'FROM',
            }),
          },
          to: {
            utc: () => ({
              format: () => 'TO',
            }),
          },
        },
        targets: [
          {
            refId: 'REFID',
            namespace: 'NAMESPACE',
            streamId: 'https://dat-b.osisoft.com/streampath',
            streamName: 'STREAM',
          },
        ],
      };
      const response = datasource.query(options as any);
      expect(backendSrv.datasourceRequest).toHaveBeenCalledWith({
        url: 'URL/ocs/streampath/data?startIndex=FROM&endIndex=TO',
        method: 'GET',
      });
      response.then((r) => {
        expect(JSON.stringify(r)).toEqual(
          JSON.stringify({
            data: [
              new MutableDataFrame({
                refId: 'REFID',
                name: 'STREAM',
                fields: [
                  {
                    name: 'TimeStamp',
                    type: FieldType.time,
                    values: [Date.parse('2020-01-01')],
                  },
                  {
                    name: 'Boolean',
                    type: FieldType.boolean,
                    values: [true],
                  },
                  {
                    name: 'Number',
                    type: FieldType.number,
                    values: [1],
                  },
                  {
                    name: 'String',
                    type: FieldType.string,
                    values: ['A'],
                  },
                ],
              }),
            ],
          })
        );
        done();
      });
    });

    it('should handle invalid query', (done) => {
      spyOn(backendSrv, 'datasourceRequest');
      const datasource = new SdsDataSource(ocsSettings, backendSrv as any);
      const options = {
        range: null,
        targets: [
          {
            refId: 'REFID',
            namespace: 'NAMESPACE',
            streamId: null,
            streamName: 'STREAM',
          },
        ],
      };
      const response = datasource.query(options as any);
      expect(backendSrv.datasourceRequest).not.toHaveBeenCalled();
      response.then((r) => {
        expect(JSON.stringify(r)).toEqual(JSON.stringify({ data: [new MutableDataFrame()] }));
        done();
      });
    });
  });

  describe('getStreams', () => {
    it('should return empty if namespace is not defined', (done) => {
      const datasource = new SdsDataSource(ocsSettings, backendSrv as any);
      datasource.namespace = '';
      const result = datasource.getStreams('');
      result.then((r) => {
        expect(r).toEqual([]);
        done();
      });
    });

    it('should query for streams', (done) => {
      const Id = 'Stream';
      spyOn(backendSrv, 'datasourceRequest').and.returnValue(Promise.resolve({ data: [{ Id }] }));
      const datasource = new SdsDataSource(ocsSettings, backendSrv as any);
      const result = datasource.getStreams('test');
      result.then((r) => {
        expect(r).toEqual([{ value: Id, label: Id }]);
        done();
      });
    });

    it('should query a community for streams', (done) => {
      const Id = 'Stream';
      const Self = 'Self';
      spyOn(backendSrv, 'datasourceRequest').and.returnValue(Promise.resolve({ data: [{ Self, Id }] }));
      const datasource = new SdsDataSource(ocsCommSettings, backendSrv as any);
      const result = datasource.getStreams('test');
      result.then((r) => {
        expect(r).toEqual([{ value: Self, label: Id }]);
        done();
      });
    });
  });

  describe('testDatasource', () => {
    it('should run a test query', (done) => {
      spyOn(backendSrv, 'datasourceRequest').and.returnValue(
        Promise.resolve({
          status: 200,
        })
      );
      const datasource = new SdsDataSource(ocsSettings, backendSrv as any);
      const response = datasource.testDatasource();
      expect(backendSrv.datasourceRequest).toHaveBeenCalledWith({
        url: 'URL/ocs/api/VERSION/tenants/TENANT/namespaces/NAMESPACE/streams',
        method: 'GET',
      });
      response.then((r) => {
        expect(r).toEqual({
          status: 'success',
          message: 'Data source is working',
        });
        done();
      });
    });

    it('should handle test failure', (done) => {
      spyOn(backendSrv, 'datasourceRequest').and.returnValue(
        Promise.resolve({
          status: 400,
          statusText: 'Error',
        })
      );
      const datasource = new SdsDataSource(ocsSettings, backendSrv as any);
      const response = datasource.testDatasource();
      expect(backendSrv.datasourceRequest).toHaveBeenCalledWith({
        url: 'URL/ocs/api/VERSION/tenants/TENANT/namespaces/NAMESPACE/streams',
        method: 'GET',
      });
      response.then((r) => {
        expect(r).toEqual({
          status: 'error',
          message: '400: Error',
        });
        done();
      });
    });
  });
});


src/DataSource.ts
import {
  DataQueryRequest,
  DataQueryResponse,
  DataSourceApi,
  DataSourceInstanceSettings,
  MutableDataFrame,
  FieldType,
  SelectableValue,
} from '@grafana/data';

import { SdsQuery, SdsDataSourceOptions, SdsDataSourceType } from 'types';

export declare type BackendSrvRequest = {
  url: string;
  method?: string;
};

export interface BackendSrv {
  datasourceRequest(options: BackendSrvRequest): Promise<any>;
}

export class SdsDataSource extends DataSourceApi<SdsQuery, SdsDataSourceOptions> {
  name: string;
  proxyUrl: string;

  type: SdsDataSourceType;
  edsPort: string;
  ocsUrl: string;
  ocsVersion: string;
  ocsTenant: string;
  ocsUseCommunity: boolean;
  ocsCommunity: string;
  oauthPassThru: boolean;
  namespace: string;

  get streamsUrl() {
    return this.type === SdsDataSourceType.OCS
      ? this.ocsUseCommunity === true
        ? `${this.proxyUrl}/ocs/api/${this.ocsVersion}/tenants/${this.ocsTenant}/search/communities/${this.ocsCommunity}/streams`
        : `${this.proxyUrl}/ocs/api/${this.ocsVersion}/tenants/${this.ocsTenant}/namespaces/${this.namespace}/streams`
      : `http://localhost:${this.edsPort}/api/v1/tenants/default/namespaces/${this.namespace}/streams`;
  }

  /** @ngInject */
  constructor(instanceSettings: DataSourceInstanceSettings<SdsDataSourceOptions>, private backendSrv: BackendSrv) {
    super(instanceSettings);
    this.name = instanceSettings.name;
    this.proxyUrl = instanceSettings.url ? instanceSettings.url.trim() : '';
    this.backendSrv = backendSrv;

    this.type = instanceSettings.jsonData?.type || SdsDataSourceType.OCS;
    this.edsPort = instanceSettings.jsonData?.edsPort || '5590';
    this.ocsUrl = instanceSettings.jsonData?.ocsUrl || '';
    this.ocsVersion = instanceSettings.jsonData?.ocsVersion || 'v1';
    this.ocsTenant = instanceSettings.jsonData?.ocsTenant || '';
    this.ocsUseCommunity = instanceSettings.jsonData?.ocsUseCommunity || false;
    this.ocsCommunity = instanceSettings.jsonData?.ocsCommunity || '';
    this.oauthPassThru = instanceSettings.jsonData?.oauthPassThru || false;
    this.namespace = instanceSettings.jsonData?.namespace || '';
  }

  async query(options: DataQueryRequest<SdsQuery>): Promise<DataQueryResponse> {
    const from = options.range?.from.utc().format();
    const to = options.range?.to.utc().format();
    const requests = options.targets.map((target) => {
      if (!target.streamId) {
        return new Promise((resolve) => resolve(null));
      }
      if (this.ocsUseCommunity) {
        const url = new URL(target.streamId);

        return this.backendSrv.datasourceRequest({
          url: `${this.proxyUrl}/ocs${url.pathname}/data?startIndex=${from}&endIndex=${to}`,
          method: 'GET',
        });
      } else {
        return this.backendSrv.datasourceRequest({
          url: `${this.streamsUrl}/${target.streamId}/data?startIndex=${from}&endIndex=${to}`,
          method: 'GET',
        });
      }
    });

    const data = await Promise.all(requests).then((responses) => {
      let i = 0;
      return responses.map((r: any) => {
        if (!r || !r.data.length) {
          return new MutableDataFrame();
        }

        const target = options.targets[i];
        i++;
        return new MutableDataFrame({
          refId: target.refId,
          name: target.streamName,
          fields: Object.keys(r.data[0]).map((name) => {
            const val0 = r.data[0][name];
            const date = Date.parse(val0);
            const num = Number(val0);
            const type =
              typeof val0 === 'string' && !isNaN(date)
                ? FieldType.time
                : val0 === true || val0 === false
                ? FieldType.boolean
                : !isNaN(num)
                ? FieldType.number
                : FieldType.string;
            return {
              name,
              values: r.data.map((d) => (type === FieldType.time ? Date.parse(d[name]) : d[name])),
              type,
            };
          }),
        });
      });
    });

    return { data };
  }

  async getStreams(query: string): Promise<Array<SelectableValue<string>>> {
    const url = query ? `${this.streamsUrl}?query=*${query}*` : this.streamsUrl;
    const requests = this.backendSrv.datasourceRequest({ url, method: 'GET' });
    if (this.ocsUseCommunity === true) {
      return await Promise.resolve(requests).then((responses) =>
        Object.keys(responses.data).map((r) => ({ value: responses.data[r].Self, label: responses.data[r].Id }))
      );
    } else if (this.namespace) {
      return await Promise.resolve(requests).then((responses) =>
        Object.keys(responses.data).map((r) => ({ value: responses.data[r].Id, label: responses.data[r].Id }))
      );
    } else {
      return await new Promise((resolve) => resolve([]));
    }
  }

  async testDatasource() {
    return this.backendSrv
      .datasourceRequest({
        url: this.streamsUrl,
        method: 'GET',
      })
      .then((r) => {
        if (r.status === 200) {
          return {
            status: 'success',
            message: 'Data source is working',
          };
        } else {
          return {
            status: 'error',
            message: `${r.status}: ${r.statusText}`,
          };
        }
      });
  }
}
